onnxruntime/python/tools/quantization/calibrate.py

```diff
@@ -226,12 +239,242 @@ def compute_range(self):
+        self.num_model_outputs = len(self.model.graph.output)
+        self.model_original_outputs = set(output.name for output in self.model.graph.output)
+        self.collector = None
+
+    def augment_graph(self):
+        '''
+        make all quantization_candidates op type nodes as part of the graph output.
+        :return: augmented ONNX model
+        '''
+        model = onnx_proto.ModelProto()
+        model.CopyFrom(self.model)
+        model = onnx.shape_inference.infer_shapes(model)
+
+        added_nodes = []
+        added_outputs = []
+        tensors, value_infos = self.select_tensors_to_calibrate(model) 
+
+        for tensor in tensors:
+            added_outputs.append(value_infos[tensor])
+
+        model.graph.node.extend(added_nodes)
+        model.graph.output.extend(added_outputs)
+        onnx.save(model, self.augmented_model_path)
+        self.augment_model = model
+
+    def clear_collected_data(self):
+        self.intermediate_outputs = []
+
+    def collect_data(self, data_reader: CalibrationDataReader):
+        '''
+        Entropy Calibrator collects operators' tensors as well as generates tensor histogram for each operator. 
+        '''
+        while True:
+            inputs = data_reader.get_next()
+            if not inputs:
+                break
+            self.intermediate_outputs.append(self.infer_session.run(None, inputs))
+
+
+        if len(self.intermediate_outputs) == 0:
+            raise ValueError("No data is collected.")
+
+        output_names = [self.infer_session.get_outputs()[i].name for i in range(len(self.intermediate_outputs[0]))]
+        output_dicts_list = [
+            dict(zip(output_names, intermediate_output)) for intermediate_output in self.intermediate_outputs
+        ]
+
+        merged_dict = {}
+        for d in output_dicts_list:
+            for k, v in d.items():
+                merged_dict.setdefault(k, []).append(v)
+
+        clean_merged_dict = dict((i, merged_dict[i]) for i in merged_dict if i not in self.model_original_outputs)
+
+        if not self.collector:
+            self.collector = HistogramCollector()
+        self.collector.collect(clean_merged_dict)
+
+    def compute_range(self):
+        ''' 
+        Compute the min-max range of tensor
+        :return: dictionary mapping: {added node names: (ReduceMin, ReduceMax) pairs }
+        '''
+        if not self.collector:
+            print("No collector created and can't generate calibration data.")
+            return None
+
+        return self.collector.get_optimal_collection_result()
+
+
+class CalibrationDataCollector(metaclass=abc.ABCMeta):
+    """
+    Base class for collecting data for calibration-based quantization.
+    """
+
+    @abc.abstractmethod
+    def collect(self, name_to_arr):
+        """
+        Generate informative data based on given data.
+            name_to_arr : dict 
+                tensor name to NDArray data 
+        """
+        raise NotImplementedError
```

Comment:
Does this abstract class need a get_optimal_collection_result method? Is this abstract class nessarary?

Reply:
Besides histogram collection, it's possible that later other collection approaches will be introduced.
It's better off to have this abstract class to make sure all the collection approaches are consistent.

get_optimal_collection_result method added.